gluten-core/src/main/scala/io/glutenproject/extension/columnar/ColumnarPullOutPreProject.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.extension.columnar
+
+import io.glutenproject.utils.{PhysicalPlanSelector, PullOutProjectHelper}
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.expressions.{Expression, NamedExpression, SortOrder}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{ProjectExec, SortExec, SparkPlan, TakeOrderedAndProjectExec}
+
+import scala.collection.mutable
+
+/**
+ * The native engine only supports executing Expressions within the project operator. When there are
+ * Expressions within physical operators such as SortExec and HashAggregateExec in SparkPlan, we
+ * need to pull out the Expressions to be executed earlier in the pre-project operator. This rule is
+ * to transform the SparkPlan at the physical plan level, constructing a SparkPlan that supports
+ * execution by the native engine.
+ */
+case class ColumnarPullOutPreProject(session: SparkSession)
```

Comment:
nit: I think it's fine to use name `PullOutPreProject`, since we are not adding another logical rule in the PR yet.